theano/tensor/nnet/ctc.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import (division, absolute_import, print_function)
+import os
+import theano.tensor as T
+from theano import config
+from theano import gof
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+from theano.tensor.extra_ops import cpu_contiguous
+from theano.gradient import grad_undefined
+
+ctc_enabled = config.ctc.enabled
+
+
+class ConnectionistTemporalClassification(gof.COp, gof.OpenMPOp):
+    """
+    CTC loss function wrapper.
+
+    Notes
+    -----
+    Using the wrapper requires that Baidu's warp-ctc library is installed and the
+    configuration variables `config.ctc.enabled` and `config.ctc.root` be properly
+    set.
+
+    """
+    __props__ = ('compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_cpu)"
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_cpu_no_grad)"
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+        gof.OpenMPOp.__init__(self)
+
+        self.compute_grad = compute_grad
+        self.costs = T.fvector(name="ctc_cost")
+        if self.compute_grad:
+            self.gradients = T.ftensor3(name="ctc_grad")
+
```

Comment:
We should use `L_op()` for CPU op also, so that we can avoid creating variables in `__init__()`. It's better for readability and code encapsulation/reusage (as an op instance may be called many times with different inputs), and we usually create variables in `make_node()` mainly!